aws-lc-rs/src/evp_pkey.rs

```diff
@@ -69,53 +57,99 @@ impl LcPtr<EVP_PKEY> {
+        let mut cbb = LcCBB::new(self.key_size_bytes() * 5);
+        if 1 != unsafe { EVP_marshal_public_key(cbb.as_mut_ptr(), *self.as_const()) } {
+            return Err(Unspecified);
+        };
+        cbb.into_vec()
+    }
 
-        let out_len = {
-            let mut cbb = LcCBB::new_fixed(<&mut [u8; PKCS8_DOCUMENT_MAX_LEN]>::try_from(
-                buffer.as_mut_slice(),
-            )?);
+    pub(crate) fn parse_rfc5280_public_key(
+        bytes: &[u8],
+        evp_pkey_type: c_int,
+    ) -> Result<Self, KeyRejected> {
+        let mut cbs = cbs::build_CBS(bytes);
+        // Also checks the validity of the key
+        let evp_pkey = LcPtr::new(unsafe { EVP_parse_public_key(&mut cbs) })
+            .map_err(|()| KeyRejected::invalid_encoding())?;
+        evp_pkey
+            .id()
+            .eq(&evp_pkey_type)
+            .then_some(evp_pkey)
+            .ok_or(KeyRejected::wrong_algorithm())
+    }
 
-            match version {
-                Version::V1 => {
-                    if 1 != unsafe { EVP_marshal_private_key(cbb.as_mut_ptr(), *self.as_const()) } {
-                        return Err(Unspecified);
-                    }
+    pub(crate) fn marshal_rfc5208_private_key(
+        &self,
+        version: Version,
+    ) -> Result<Vec<u8>, Unspecified> {
+        let key_size_bytes = TryInto::<usize>::try_into(unsafe { EVP_PKEY_bits(*self.as_const()) })
+            .expect("fit in usize")
+            / 8;
+        let mut cbb = LcCBB::new(key_size_bytes * 5);
+        match version {
+            Version::V1 => {
+                if 1 != unsafe { EVP_marshal_private_key(cbb.as_mut_ptr(), *self.as_const()) } {
+                    return Err(Unspecified);
                 }
-                Version::V2 => {
-                    if 1 != unsafe {
-                        EVP_marshal_private_key_v2(cbb.as_mut_ptr(), *self.as_const())
-                    } {
-                        return Err(Unspecified);
-                    }
+            }
+            Version::V2 => {
+                if 1 != unsafe { EVP_marshal_private_key_v2(cbb.as_mut_ptr(), *self.as_const()) } {
+                    return Err(Unspecified);
                 }
             }
-            cbb.finish()?
-        };
-
-        buffer.truncate(out_len);
+        }
+        cbb.into_vec()
+    }
 
-        Ok(buffer.into_boxed_slice())
+    pub(crate) fn parse_rfc5208_private_key(
+        bytes: &[u8],
+        evp_pkey_type: c_int,
+    ) -> Result<Self, KeyRejected> {
+        let mut cbs = cbs::build_CBS(bytes);
+        // Also checks the validity of the key
+        let evp_pkey = LcPtr::new(unsafe { EVP_parse_private_key(&mut cbs) })
+            .map_err(|()| KeyRejected::invalid_encoding())?;
+        evp_pkey
+            .id()
+            .eq(&evp_pkey_type)
+            .then_some(evp_pkey)
+            .ok_or(KeyRejected::wrong_algorithm())
     }
```

Comment:
Should we consider following the same pattern as `mod sec` and `mod rfc5915` and move these functions to another `mod rfc5208` in `encoding.rs`? It seems like they belong in the same realm at first glance, but I do see how this makes things handy for `EVP_PKEY`.